jib-core/src/main/java/com/google/cloud/tools/jib/filesystem/XdgDirectories.java

```diff
@@ -0,0 +1,174 @@
+
+    } else if (osName.contains("mac") || osName.contains("darwin")) {
+      // Use '~/Library/Application Support/' for macOS.
+      Path applicationSupport = Paths.get(userHome, "Library", "Application Support");
+      if (!Files.exists(applicationSupport)) {
+        LOGGER.warning(applicationSupport + " does not exist");
+        return xdgPath;
+      }
+      return applicationSupport;
+    }
+
+    throw new IllegalStateException("Unknown OS: " + rawOsName);
+  }
+
+  /**
+   * Returns config directory based on {@code $XDG_CONFIG_HOME}, if available, or resolves the
+   * OS-specific user config directory.
+   *
+   * <p>For Linux, this is {@code $HOME/.config/google-cloud-tools-java/jib/}.
+   *
+   * <p>For Windows, this is {@code %LOCALAPPDATA%\Google\Jib/}.
+   *
+   * <p>For macOS, this is {@code $HOME/Library/Preferences/Google/Jib/}.
+   */
+  @VisibleForTesting
+  static Path getConfigHome(Properties properties, Map<String, String> environment) {
+    Path windowsSubDirectory = JIB_SUBDIRECTORY_OTHER.resolve("Config");
+    String rawOsName = properties.getProperty("os.name");
+    String osName = rawOsName.toLowerCase(Locale.ENGLISH);
+
+    // Use environment variable $XDG_CONFIG_HOME if set and not empty.
+    String xdgConfigHome = environment.get("XDG_CONFIG_HOME");
+    if (xdgConfigHome != null && !xdgConfigHome.trim().isEmpty()) {
+      if (osName.contains("linux")) {
+        return Paths.get(xdgConfigHome).resolve(JIB_SUBDIRECTORY_LINUX);
+      } else if (osName.contains("windows")) {
+        return Paths.get(xdgConfigHome).resolve(windowsSubDirectory);
+      }
+      return Paths.get(xdgConfigHome).resolve(JIB_SUBDIRECTORY_OTHER);
+    }
+
+    String userHome = properties.getProperty("user.home");
+    Path xdgPath = Paths.get(userHome, ".config");
+
+    if (osName.contains("linux")) {
+      return xdgPath.resolve(JIB_SUBDIRECTORY_LINUX);
+
+    } else if (osName.contains("windows")) {
+      // Use %LOCALAPPDATA% for Windows.
+      String localAppDataEnv = environment.get("LOCALAPPDATA");
+      if (localAppDataEnv == null || localAppDataEnv.trim().isEmpty()) {
+        LOGGER.warning("LOCALAPPDATA environment is invalid or missing");
+        return xdgPath.resolve(windowsSubDirectory);
+      }
+      Path localAppData = Paths.get(localAppDataEnv);
+      if (!Files.exists(localAppData)) {
+        LOGGER.warning(localAppData + " does not exist");
+        return xdgPath.resolve(windowsSubDirectory);
+      }
+      return localAppData.resolve(windowsSubDirectory);
+
+    } else if (osName.contains("mac") || osName.contains("darwin")) {
+      // Use '~/Library/Preferences/' for macOS.
+      Path applicationSupport = Paths.get(userHome, "Library", "Preferences");
+      if (!Files.exists(applicationSupport)) {
+        LOGGER.warning(applicationSupport + " does not exist");
+        return xdgPath.resolve(JIB_SUBDIRECTORY_OTHER);
+      }
+      return applicationSupport.resolve(JIB_SUBDIRECTORY_OTHER);
```

Comment:
Looks like this structure is being repeated.
```
if (linux) {
   return path.resolve(JIB_SUBDIRECTORY_LINUX);
} else if (windows) {
   return path.resolve(windowsSubDirectory);
} else /* if (mac) {
   return path.resolve(JIB_SUBDIRECTORY_OTHER);
}
```
Perhaps this logic can be a separate method, like `getConfigHome(getBaseConfigHome(...))` (whatever suitable method names)? If this is indeed repeated, I think factoring this out will make the code more readable.

Reply:
Sure, I'll give it a try.